File: sdk/python_sdk/lemon_kami.py
"""
Lemon Kami Python SDK
卡密验证客户端 SDK
"""

import json
import time
import uuid
import hashlib
import hmac
import platform
from typing import Optional, Dict, Any
import requests
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_v1_5, AES
from Crypto.Util.Padding import pad, unpad
import base64
import os


class LemonKamiSDK:
    """小柠檬网络验证 SDK"""
    
    def __init__(
        self,
        app_id: str,
        app_secret: str,
        server_url: str = "http://localhost:8000",
        rsa_public_key: Optional[str] = None
    ):
        """
        初始化 SDK
        
        Args:
            app_id: 应用ID
            app_secret: 应用密钥
            server_url: 服务器地址
            rsa_public_key: RSA公钥（可选，会自动从服务器获取）
        """
        self.app_id = app_id
        self.app_secret = app_secret
        self.server_url = server_url.rstrip('/')
        self.rsa_public_key = rsa_public_key
        # 移除设备UUID缓存，每次重新生成设备指纹
        self.fingerprint = self._generate_device_fingerprint()
        
        # 如果未提供公钥，自动获取
        if not self.rsa_public_key:
            self._fetch_public_key()
    
    def _generate_device_fingerprint(self) -> str:
        """
        生成设备指纹（每次都重新计算，不依赖缓存）
        
        使用多种硬件信息组合生成唯一的设备标识符，确保每次执行结果一致且难以伪造
        """
        # 收集稳定的硬件信息
        info = {
            # CPU 信息（最稳定）
            "machine": platform.machine(),
            "processor": platform.processor(),
            # 系统节点名（计算机名）
            "node": platform.node(),
            # 操作系统平台信息
            "platform": platform.platform(),
            # 网络适配器 MAC 地址（最稳定，但需要权限）
            "mac_addresses": self._get_mac_addresses(),
            # 主板序列号或其他唯一标识（如果可用）
            "system_info": self._get_system_unique_id(),
        }
        
        # 生成指纹哈希 - 使用 SHA256 确保一致性
        fingerprint_str = json.dumps(info, sort_keys=True)
        return hashlib.sha256(fingerprint_str.encode()).hexdigest()
    
    def _get_system_unique_id(self) -> str:
        """
        获取系统唯一标识符（跨平台兼容）
        
        Returns:
            系统唯一标识符字符串
        """
        try:
            system = platform.system()
            
            if system == "Windows":
                # Windows: 尝试获取机器GUID或产品ID
                import subprocess
                try:
                    # 尝试获取 MachineGuid
                    result = subprocess.run(
                        ['reg', 'query', 'HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Cryptography', '/v', 'MachineGuid'],
                        capture_output=True, text=True, timeout=5
                    )
                    if result.returncode == 0:
                        for line in result.stdout.split('\n'):
                            if 'MachineGuid' in line:
                                return line.strip().split()[-1]
                except:
                    pass
                    
                try:
                    # 备选方案：获取产品ID
                    result = subprocess.run(
                        ['wmic', 'csproduct', 'get', 'UUID'],
                        capture_output=True, text=True, timeout=5
                    )
                    lines = result.stdout.strip().split('\n')
                    if len(lines) > 1:
                        uuid_val = lines[1].strip()
                        if uuid_val and uuid_val != 'UUID':
                            return uuid_val
                except:
                    pass
                    
            elif system == "Darwin":  # macOS
                # macOS: 获取硬件UUID
                import subprocess
                try:
                    result = subprocess.run(
                        ['system_profiler', 'SPHardwareDataType'],
                        capture_output=True, text=True, timeout=5
                    )
                    for line in result.stdout.split('\n'):
                        if 'Hardware UUID' in line:
                            return line.split(':')[1].strip()
                except:
                    pass
                    
            else:  # Linux
                # Linux: 尝试获取 machine-id
                try:
                    with open('/etc/machine-id', 'r') as f:
                        return f.read().strip()
                except:
                    pass
                    
                # 备选：获取 dbus machine id
                try:
                    with open('/var/lib/dbus/machine-id', 'r') as f:
                        return f.read().strip()
                except:
                    pass
        except:
            pass
        
        # 如果所有方法都失败，返回一个基于平台信息的标识符
        return f"{platform.system()}_{platform.release()}_{platform.machine()}"
    
    def _get_mac_addresses(self) -> list:
        """
        获取网络适配器的 MAC 地址
        
        Returns:
            MAC 地址列表
        """
        import re
        macs = []
        try:
            # Windows
            if platform.system() == "Windows":
                import subprocess
                result = subprocess.run(['getmac'], capture_output=True, text=True)
                # 提取 MAC 地址
                mac_pattern = r'([0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2})'
                macs = re.findall(mac_pattern, result.stdout)
            else:
                # Linux/Mac
                import subprocess
                result = subprocess.run(['ifconfig'], capture_output=True, text=True)
                mac_pattern = r'([0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2}[:-][0-9A-Fa-f]{2})'
                macs = re.findall(mac_pattern, result.stdout)
        except:
            pass
        
        # 去重并排序，确保顺序一致
        return sorted(list(set(macs)))
    
    def _fetch_public_key(self):
        """从服务器获取RSA公钥"""
        try:
            response = requests.get(
                f"{self.server_url}/api/v1/sdk/public-key",
                params={"app_id": self.app_id},
                timeout=10
            )
            
            if response.status_code == 200:
                data = response.json()
                self.rsa_public_key = data.get("public_key")
                
                # 验证公钥是否成功获取
                if not self.rsa_public_key:
                    raise Exception(f"服务器返回的公钥为空: {data}")
            else:
                raise Exception(f"HTTP {response.status_code}: {response.text}")
                
        except requests.exceptions.ConnectionError:
            raise Exception(f"无法连接到服务器: {self.server_url}，请检查服务器是否运行")
        except requests.exceptions.Timeout:
            raise Exception(f"请求超时: {self.server_url}")
        except Exception as e:
            if "获取公钥失败" in str(e):
                raise
            raise Exception(f"获取公钥失败: {e}")
    
    def _encrypt_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        加密数据
        
        使用 RSA + AES 混合加密：
        1. 生成随机 AES 密钥
        2. 用 AES 加密业务数据
        3. 用 RSA 加密 AES 密钥
        4. 返回符合后端格式的字典
        """
        # 验证公钥是否存在
        if not self.rsa_public_key:
            raise Exception("RSA 公钥未初始化，请检查服务器连接或手动指定公钥")
        
        # 生成随机 AES 密钥和 IV (使用 AES-128, 16字节)
        aes_key = os.urandom(16)  # AES-128
        aes_iv = os.urandom(16)
        
        # AES 加密业务数据
        data_json = json.dumps(data, ensure_ascii=False).encode('utf-8')
        cipher_aes = AES.new(aes_key, AES.MODE_CBC, aes_iv)
        encrypted_data = cipher_aes.encrypt(pad(data_json, AES.block_size))
        
        # RSA 加密 AES 密钥
        # 先将 AES 密钥转为 Base64 字符串，再 RSA 加密（与 JS SDK 保持一致）
        aes_key_base64 = base64.b64encode(aes_key).decode('utf-8')
        rsa_key = RSA.import_key(self.rsa_public_key)
        cipher_rsa = PKCS1_v1_5.new(rsa_key)
        encrypted_aes_key = cipher_rsa.encrypt(aes_key_base64.encode('utf-8'))
        
        # 生成时间戳和随机数
        timestamp = int(time.time())
        nonce = hashlib.md5(f"{time.time()}{os.urandom(8).hex()}".encode()).hexdigest()[:16]
        
        # 生成 HMAC-SHA256 签名（与后端保持一致）
        sign_str = f"{timestamp}{nonce}{base64.b64encode(encrypted_data).decode()}"
        signature = hmac.new(
            self.app_secret.encode('utf-8'),
            sign_str.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()
        
        # 返回符合后端格式的字典
        return {
            "app_id": self.app_id,
            "timestamp": timestamp,
            "nonce": nonce,
            "sign": signature,
            "encrypted_key": base64.b64encode(encrypted_aes_key).decode(),
            "encrypted_data": base64.b64encode(encrypted_data).decode(),
            "iv": base64.b64encode(aes_iv).decode()
        }
    
    def verify(self, kami_code: str) -> Dict[str, Any]:
        """
        验证卡密
        
        Args:
            kami_code: 卡密代码
            
        Returns:
            验证结果字典
        """
        # 构建请求数据（不再使用 device_uuid，只使用动态生成的 fingerprint）
        request_data = {
            "kami": kami_code,
            "fingerprint": self.fingerprint,
            "_app_info": {
                "app_id": self.app_id
            }
        }
        
        # 加密数据
        encrypted_payload = self._encrypt_data(request_data)
        
        # 发送请求
        try:
            response = requests.post(
                f"{self.server_url}/api/v1/sdk/verify",
                json=encrypted_payload,
                headers={"Content-Type": "application/json"},
                timeout=10
            )
            if response.status_code == 200:
                result = response.json()
                # 检查是否是加密响应
                if result.get("encrypted"):
                    return self._decrypt_response(result)
                return result
            else:
                return {
                    "success": False,
                    "message": f"HTTP {response.status_code}: {response.text}"
                }
        except Exception as e:
            return {
                "success": False,
                "message": f"网络错误: {str(e)}"
            }
    
    def report_event(
        self,
        kami_code: str,
        event_type: str,
        extra_data: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        上报行为事件
        
        Args:
            kami_code: 卡密代码
            event_type: 事件类型（如：login, level_up, purchase等）
            extra_data: 额外数据
            
        Returns:
            上报结果
        """
        request_data = {
            "kami": kami_code,
            "event_type": event_type,
            "extra_data": extra_data or {},
            "_app_info": {
                "app_id": self.app_id
            }
        }
        
        encrypted_payload = self._encrypt_data(request_data)
        
        try:
            response = requests.post(
                f"{self.server_url}/api/v1/sdk/report",
                json=encrypted_payload,
                headers={"Content-Type": "application/json"},
                timeout=10
            )
            
            if response.status_code == 200:
                result = response.json()
                # 检查是否是加密响应
                if result.get("encrypted"):
                    return self._decrypt_response(result)
                return result
            else:
                return {
                    "success": False,
                    "message": f"HTTP {response.status_code}: {response.text}"
                }
        except Exception as e:
            return {
                "success": False,
                "message": f"网络错误: {str(e)}"
            }
    
    def heartbeat(self, kami_code: str) -> Dict[str, Any]:
        """
        发送心跳
        
        Args:
            kami_code: 卡密代码
            
        Returns:
            心跳结果
        """
        return self.report_event(kami_code, "heartbeat")

    def release_device(self, kami_code: str) -> Dict[str, Any]:
        """
        释放当前设备占用名额。

        建议在软件正常退出或用户退出登录时调用，用于一机多码/多设备限制场景。
        """
        request_data = {
            "kami": kami_code,
            "fingerprint": self.fingerprint,
            "_app_info": {
                "app_id": self.app_id
            }
        }

        encrypted_payload = self._encrypt_data(request_data)

        try:
            response = requests.post(
                f"{self.server_url}/api/v1/sdk/release-device",
                json=encrypted_payload,
                headers={"Content-Type": "application/json"},
                timeout=10
            )

            if response.status_code == 200:
                result = response.json()
                if result.get("encrypted"):
                    return self._decrypt_response(result)
                return result
            return {
                "success": False,
                "message": f"HTTP {response.status_code}: {response.text}"
            }
        except Exception as e:
            return {
                "success": False,
                "message": f"网络错误: {str(e)}"
            }
    
    def _decrypt_response(self, encrypted_response: Dict[str, Any]) -> Dict[str, Any]:
        """
        解密响应数据
        
        Args:
            encrypted_response: 加密的响应数据
            
        Returns:
            解密后的数据字典
        """
        import hmac as hmac_module
        
        try:
            # 提取加密字段
            timestamp = encrypted_response.get("timestamp")
            nonce = encrypted_response.get("nonce")
            sign = encrypted_response.get("sign")
            encrypted_key_b64 = encrypted_response.get("encrypted_key")
            encrypted_data_b64 = encrypted_response.get("encrypted_data")
            iv_b64 = encrypted_response.get("iv")
            
            if not all([timestamp, nonce, sign, encrypted_key_b64, encrypted_data_b64, iv_b64]):
                raise Exception("响应数据格式错误")
            
            # 验证签名
            sign_str = f"{timestamp}{nonce}{encrypted_data_b64}"
            expected_sign = hmac_module.new(
                self.app_secret.encode('utf-8'),
                sign_str.encode('utf-8'),
                hashlib.sha256
            ).hexdigest()
            
            if not hmac_module.compare_digest(expected_sign, sign):
                raise Exception("响应签名验证失败")
            
            # AES 密钥已经是明文（Base64编码）
            aes_key = base64.b64decode(encrypted_key_b64)
            aes_iv = base64.b64decode(iv_b64)
            encrypted_data = base64.b64decode(encrypted_data_b64)
            
            # AES 解密
            cipher = AES.new(aes_key, AES.MODE_CBC, aes_iv)
            decrypted_padded = cipher.decrypt(encrypted_data)
            
            # 去除填充
            from Crypto.Util.Padding import unpad
            decrypted_data = unpad(decrypted_padded, AES.block_size)
            
            # 解析 JSON
            return json.loads(decrypted_data.decode('utf-8'))
            
        except Exception as e:
            return {
                "success": False,
                "message": f"响应解密失败: {str(e)}"
            }

File: sdk/js_sdk/lemon-kami.js
/**
 * Lemon Kami JavaScript SDK
 * 卡密验证客户端 SDK (使用 crypto-js + JSEncrypt)
 */

class LemonKamiSDK {
  /**
   * 初始化 SDK
   * @param {Object} config - 配置对象
   * @param {string} config.appId - 应用ID
   * @param {string} config.appSecret - 应用密钥
   * @param {string} [config.serverUrl='http://localhost:8000'] - 服务器地址
   * @param {string} [config.rsaPublicKey] - RSA公钥（可选）
   * @param {boolean} [config.autoReleaseOnUnload=true] - 页面关闭时自动释放设备名额
   */
  constructor(config) {
    // 检查依赖库是否已加载
    if (typeof CryptoJS === 'undefined') {
      throw new Error('CryptoJS 未加载，请先引入 crypto-js 库');
    }
    if (typeof JSEncrypt === 'undefined') {
      throw new Error('JSEncrypt 未加载，请先引入 jsencrypt 库');
    }
    
    this.appId = config.appId;
    this.appSecret = config.appSecret;
    this.serverUrl = (config.serverUrl || 'http://localhost:8000').replace(/\/$/, '');
    this.rsaPublicKey = config.rsaPublicKey || null;
    this.deviceUuid = this._getDeviceUuid();
    this.fingerprint = ''; // 初始化为空字符串
    this.currentKamiCode = null;
    this.autoReleaseOnUnload = config.autoReleaseOnUnload !== false;
    this._releaseOnUnloadHandler = () => {
      if (this.autoReleaseOnUnload && this.currentKamiCode) {
        this.releaseDevice(this.currentKamiCode, { keepalive: true });
      }
    };

    // 异步生成指纹
    this._fingerprintReady = this._initFingerprint();

    // 如果未提供公钥，自动获取
    if (!this.rsaPublicKey) {
      this._fetchPublicKey();
    }

    if (typeof window !== 'undefined' && window.addEventListener) {
      window.addEventListener('pagehide', this._releaseOnUnloadHandler);
      window.addEventListener('beforeunload', this._releaseOnUnloadHandler);
    }
  }

  /**
   * 异步初始化设备指纹
   * @private
   */
  async _initFingerprint() {
    this.fingerprint = await this._generateFingerprint();
  }

  /**
   * 获取设备UUID
   * @returns {string} 设备UUID
   * @private
   */
  _getDeviceUuid() {
    let uuid = localStorage.getItem('device_uuid');
    if (!uuid) {
      uuid = this._generateUuid();
      localStorage.setItem('device_uuid', uuid);
    }
    return uuid;
  }

  /**
   * 生成UUID
   * @returns {string} UUID
   * @private
   */
  _generateUuid() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      const r = Math.random() * 16 | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
  }

  /**
   * 生成设备指纹
   * @returns {string} 设备指纹哈希
   * @private
   */
  _generateFingerprint() {
    const info = {
      platform: navigator.platform,
      userAgent: navigator.userAgent,
      language: navigator.language,
      screenResolution: `${screen.width}x${screen.height}`,
      colorDepth: screen.colorDepth,
      timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
    };

    const fingerprintStr = JSON.stringify(info);
    return this._sha256(fingerprintStr);
  }

  /**
   * SHA256 哈希
   * @param {string} str - 输入字符串
   * @returns {string} SHA256 哈希值
   * @private
   */
  _sha256(str) {
    return CryptoJS.SHA256(str).toString(CryptoJS.enc.Hex);
  }

  /**
   * HMAC-SHA256 签名
   * @param {string} str - 输入字符串
   * @param {string} secret - 密钥
   * @returns {string} HMAC-SHA256 签名
   * @private
   */
  _hmacSha256(str, secret) {
    return CryptoJS.HmacSHA256(str, secret).toString(CryptoJS.enc.Hex);
  }

  /**
   * 从服务器获取RSA公钥
   * @private
   */
  async _fetchPublicKey() {
    try {
      const response = await fetch(
        `${this.serverUrl}/api/v1/sdk/public-key?app_id=${this.appId}`
      );
      if (response.ok) {
        const data = await response.json();
        this.rsaPublicKey = data.public_key;
      }
    } catch (error) {
      // 静默失败，公钥将在需要时重新获取
    }
  }

  /**
   * 加密数据
   * @param {Object} data - 要加密的数据
   * @returns {Promise<Object>} 加密后的数据对象
   * @private
   */
  async _encryptData(data) {
    // 验证公钥是否存在
    if (!this.rsaPublicKey) {
      throw new Error("RSA 公钥未初始化，请检查服务器连接或手动指定公钥");
    }

    // 生成随机 AES 密钥和 IV (使用 AES-128, 16字节)
    const aesKey = CryptoJS.lib.WordArray.random(16);
    const aesIv = CryptoJS.lib.WordArray.random(16);

    // AES 加密业务数据
    const dataJson = JSON.stringify(data);
    const encryptedData = this._aesEncrypt(dataJson, aesKey, aesIv);

    // RSA 加密 AES 密钥
    // 将 WordArray 转换为原始字节数组，然后转为字符串（每个字节一个字符）
    const aesKeyBytes = new Uint8Array(aesKey.sigBytes);
    for (let i = 0; i < aesKey.sigBytes; i++) {
      aesKeyBytes[i] = (aesKey.words[i >>> 2] >>> (24 - (i % 4) * 8)) & 0xff;
    }
    
    // 将字节数组转为字符串（每个字节作为一个字符）
    let aesKeyStr = '';
    for (let i = 0; i < aesKeyBytes.length; i++) {
      aesKeyStr += String.fromCharCode(aesKeyBytes[i]);
    }
    
    const encryptedAesKey = this._rsaEncrypt(aesKeyStr);

    // 生成时间戳和随机数
    const timestamp = Math.floor(Date.now() / 1000);
    const nonce = Math.random().toString(36).substring(2, 18);

    // 生成 HMAC-SHA256 签名（与后端保持一致）
    const encryptedDataBase64 = encryptedData.toString(CryptoJS.enc.Base64);
    const signStr = `${timestamp}${nonce}${encryptedDataBase64}`;
    const signature = this._hmacSha256(signStr, this.appSecret);

    // 返回符合后端格式的字典
    const result = {
      app_id: this.appId,
      timestamp: timestamp,
      nonce: nonce,
      sign: signature,
      encrypted_key: encryptedAesKey,
      encrypted_data: encryptedDataBase64,
      iv: aesIv.toString(CryptoJS.enc.Base64)
    };

    return result;
  }

  /**
   * AES 加密
   * @param {string} data - 数据
   * @param {WordArray} key - 密钥
   * @param {WordArray} iv - 初始化向量
   * @returns {WordArray} 加密后的数据
   * @private
   */
  _aesEncrypt(data, key, iv) {
    const encrypted = CryptoJS.AES.encrypt(data, key, {
      iv: iv,
      mode: CryptoJS.mode.CBC,
      padding: CryptoJS.pad.Pkcs7
    });
    
    return encrypted.ciphertext;
  }

  /**
   * RSA 加密（使用 JSEncrypt）
   * @param {string} data - 原始字符串（必须是 Latin1 编码）
   * @returns {string} 加密后的数据（Base64）
   * @private
   */
  _rsaEncrypt(data) {
    // JSEncrypt 内部会将字符串转为 UTF-8，导致字节膨胀
    // 所以我们需要先将字符串转为字节数组，然后手动处理
    
    const encrypt = new JSEncrypt();
    encrypt.setPublicKey(this.rsaPublicKey);
    
    // 将字符串转为字节数组（每个字符一个字节）
    const bytes = new Uint8Array(data.length);
    for (let i = 0; i < data.length; i++) {
      bytes[i] = data.charCodeAt(i) & 0xFF;
    }
    
    // 将字节数组转为 Base64
    let binary = '';
    for (let i = 0; i < bytes.length; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    const base64Data = btoa(binary);
    
    // 加密 Base64 字符串
    const encrypted = encrypt.encrypt(base64Data);
    
    if (!encrypted) {
      throw new Error('RSA 加密失败');
    }
    
    return encrypted;
  }

  /**
   * 验证卡密
   * @param {string} kamiCode - 卡密代码
   * @returns {Promise<Object>} 验证结果
   */
  async verify(kamiCode) {
    await this._fingerprintReady;
    const requestData = {
      kami: kamiCode,
      uuid: this.deviceUuid,
      fingerprint: this.fingerprint,
      _app_info: {
        app_id: this.appId
      }
    };

    const encryptedData = await this._encryptData(requestData);

    try {
      const response = await fetch(`${this.serverUrl}/api/v1/sdk/verify`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(encryptedData)
      });

      const result = await response.json();

      // 检查是否是加密响应
      const finalResult = result.encrypted ? await this._decryptResponse(result) : result;
      if (finalResult.success) {
        this.currentKamiCode = kamiCode;
      }
      return finalResult;
    } catch (error) {
      return {
        success: false,
        message: `网络错误: ${error.message}`
      };
    }
  }

  /**
   * 释放当前设备占用名额
   * @param {string} [kamiCode] - 卡密代码，默认使用最近验证成功的卡密
   * @param {Object} [options={}] - 请求选项
   * @param {boolean} [options.keepalive=false] - 页面关闭时使用 keepalive 请求
   * @returns {Promise<Object>} 释放结果
   */
  async releaseDevice(kamiCode = this.currentKamiCode, options = {}) {
    if (!kamiCode) {
      return {
        success: false,
        message: '暂无需要释放的卡密'
      };
    }

    await this._fingerprintReady;
    const requestData = {
      kami: kamiCode,
      uuid: this.deviceUuid,
      fingerprint: this.fingerprint,
      _app_info: {
        app_id: this.appId
      }
    };

    try {
      const encryptedData = await this._encryptData(requestData);
      const response = await fetch(`${this.serverUrl}/api/v1/sdk/release-device`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(encryptedData),
        keepalive: Boolean(options.keepalive)
      });

      const result = await response.json();
      const finalResult = result.encrypted ? await this._decryptResponse(result) : result;
      if (finalResult.success) {
        this.currentKamiCode = null;
      }
      return finalResult;
    } catch (error) {
      return {
        success: false,
        message: `释放设备失败: ${error.message}`
      };
    }
  }

  /**
   * 上报行为事件
   * @param {string} kamiCode - 卡密代码
   * @param {string} eventType - 事件类型
   * @param {Object} [extraData={}] - 额外数据
   * @returns {Promise<Object>} 上报结果
   */
  async reportEvent(kamiCode, eventType, extraData = {}) {
    const requestData = {
      kami: kamiCode,
      event_type: eventType,
      extra_data: extraData,
      _app_info: {
        app_id: this.appId
      }
    };

    const encryptedData = await this._encryptData(requestData);

    try {
      const response = await fetch(`${this.serverUrl}/api/v1/sdk/report`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(encryptedData)
      });

      const result = await response.json();

      // 检查是否是加密响应
      if (result.encrypted) {
        return await this._decryptResponse(result);
      }

      return result;
    } catch (error) {
      return {
        success: false,
        message: `网络错误: ${error.message}`
      };
    }
  }

  /**
   * 发送心跳
   * @param {string} kamiCode - 卡密代码
   * @returns {Promise<Object>} 心跳结果
   */
  heartbeat(kamiCode) {
    return this.reportEvent(kamiCode, 'heartbeat');
  }

  /**
   * 销毁 SDK 实例，释放当前设备名额并移除自动释放监听
   * @returns {Promise<Object>} 释放结果
   */
  async destroy() {
    if (typeof window !== 'undefined' && window.removeEventListener) {
      window.removeEventListener('pagehide', this._releaseOnUnloadHandler);
      window.removeEventListener('beforeunload', this._releaseOnUnloadHandler);
    }
    return await this.releaseDevice(this.currentKamiCode);
  }

  /**
   * 解密响应数据
   * @param {Object} encryptedResponse - 加密的响应数据
   * @returns {Promise<Object>} 解密后的数据
   * @private
   */
  async _decryptResponse(encryptedResponse) {
    try {
      // 提取加密字段
      const { timestamp, nonce, sign, encrypted_key, encrypted_data, iv } = encryptedResponse;

      if (!timestamp || !nonce || !sign || !encrypted_key || !encrypted_data || !iv) {
        throw new Error("响应数据格式错误");
      }

      // 验证签名
      const signStr = `${timestamp}${nonce}${encrypted_data}`;
      const expectedSign = this._hmacSha256(signStr, this.appSecret);

      if (expectedSign !== sign) {
        throw new Error("响应签名验证失败");
      }

      // 后端返回的 encrypted_key 只是 Base64 编码，不是 RSA 加密
      // 直接 Base64 解码得到 AES 密钥
      const aesKey = CryptoJS.enc.Base64.parse(encrypted_key);
      const aesIv = CryptoJS.enc.Base64.parse(iv);
      const encryptedData = CryptoJS.enc.Base64.parse(encrypted_data);

      // AES 解密
      const decrypted = CryptoJS.AES.decrypt(
        { ciphertext: encryptedData },
        aesKey,
        {
          iv: aesIv,
          mode: CryptoJS.mode.CBC,
          padding: CryptoJS.pad.Pkcs7
        }
      );

      // 转换为字符串
      const decryptedText = decrypted.toString(CryptoJS.enc.Utf8);

      if (!decryptedText) {
        throw new Error("解密失败");
      }

      // 解析 JSON
      return JSON.parse(decryptedText);

    } catch (error) {
      return {
        success: false,
        message: `响应解密失败: ${error.message}`
      };
    }
  }
}

// Export for CommonJS and browser globals.
if (typeof module !== 'undefined' && module.exports) {
  module.exports = LemonKamiSDK;
  module.exports.LemonKamiSDK = LemonKamiSDK;
} else {
  window.LemonKamiSDK = LemonKamiSDK;
}
